crates/zizmor/src/audit/undocumented_permissions.rs

```diff
@@ -0,0 +1,153 @@
+        let Permissions::Explicit(perms) = permissions else {
+            return Ok(None);
+        };
+
+        if perms.is_empty() {
+            return Ok(None);
+        }
+
+        // Skip if it only contains "contents: read" which is common and self-explanatory
+        if perms.len() == 1
+            && perms
+                .get("contents")
+                .is_some_and(|p| *p == Permission::Read)
+        {
+            return Ok(None);
+        }
+
+        // Check each individual permission for documentation
+        let mut undocumented_permissions = Vec::new();
+        let base_location = location.clone().primary();
+
+        for (perm_name, _perm_value) in perms {
+            let individual_perm_location = base_location
+                .clone()
+                .with_keys(["permissions".into(), perm_name.as_str().into()]);
+
+            if !self.has_explanatory_comment(&individual_perm_location, workflow) {
+                undocumented_permissions.push(perm_name.as_str());
+            }
+        }
+
+        // Only create a finding if there are actually undocumented permissions
+        if !undocumented_permissions.is_empty() {
+            let perm_location = base_location.with_keys(["permissions".into()]);
+
+            Ok(Some(
+                Self::finding()
+                    .severity(Severity::Low)
+                    .confidence(Confidence::High)
+                    .persona(Persona::Pedantic)
+                    .add_location(perm_location.annotated(
+                        "consider adding comments to document the purpose of each permission",
+                    ))
+                    .build(workflow)?,
+            ))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn has_explanatory_comment(
+        &self,
+        location: &SymbolicLocation,
+        workflow: &crate::models::workflow::Workflow,
+    ) -> bool {
+        let document = workflow.as_document();
+
+        // Try to get the feature for this permission location
+        let Ok(feature) = document.query_pretty(&location.route) else {
+            // If we can't find the feature, assume it's undocumented to be safe.
+            // This handles rare edge cases like malformed routes or yamlpath internal errors.
+            return false;
+        };
+
+        // Get comments for this feature - yamlpath already extracts them!
+        let comments = document.feature_comments(&feature);
+
+        // Check if there are any meaningful comments
+        for comment_feature in &comments {
+            let comment_text = document.extract(comment_feature);
+            // Remove the '#' and trim whitespace
+            let comment_content = comment_text
+                .strip_prefix('#')
+                .unwrap_or(comment_text)
+                .trim();
+
+            // Require a non-empty comment
+            if !comment_content.is_empty() {
+                return true;
+            }
+        }
```

Comment:
I'd recommend using the `SymbolicLocation::concretize` API here -- that'll give you a `Location` that contains a `concrete: Feature` which in turn has a `comments` field.

(That API is basically identical internally to what's written here, but it's pre-composed and well covered by existing tests.)